DataFormats/Common/interface/RefToBase.h

```diff
@@ -228,43 +233,20 @@ namespace edm {
     return holder_->key();
   }
 
-  namespace {
-    // If the template parameters are classes or structs they should be
-    // related by inheritance, otherwise they should be the same type.
-    template <typename T, typename U>
-    typename std::enable_if<std::is_class<T>::value>::type checkTypeCompatibility() {
-      static_assert(std::is_base_of<T, U>::value || std::is_base_of<U, T>::value,
-                    "RefToBase::castTo error element types are not related by inheritance");
```

Comment:
How clear is the error message from the concepts compared to this `static_assert` message?

(on the other hand, I'm not sure how much we care in the ease-of-use for new uses given that we deprecated `RefToBase` https://github.com/cms-sw/cmssw/issues/42737)

Reply:
So the new error message from gcc is (minus the extra coloring it does)
```
src/DataFormats/Common/test/reftobase_t.cpp:70:45: error: no matching function for call to 'edm::RefToBase<testreftobase::Base>::castTo<edm::Ref<std::vector<testreftobase::BadType, std::allocator<testreftobase::BadType> >, testreftobase::BadType, edm::refhelper::FindUsingAdvance<std::vector<testreftobase::BadType, std::allocator<testreftobase::BadType> >, testreftobase::BadType> > >()'
   70 |       b1.castTo<Ref<std::vector<BadType> > >();
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~
In file included from src/DataFormats/Common/test/reftobase_t.cpp:3:
src/DataFormats/Common/interface/RefToBase.h:107:9: note: candidate: 'template<class REF>  requires requires{typename REF::value_type;requires(is_same_v<T, typename REF::value_type>) || (is_base_of_v<T, typename REF::value_type>) || (is_base_of_v<typename REF::value_type, T>);} REF edm::RefToBase<T>::castTo() const [with T = testreftobase::Base]'
  107 |     REF castTo() const;
      |         ^~~~~~
src/DataFormats/Common/interface/RefToBase.h:107:9: note:   template argument deduction/substitution failed:
src/DataFormats/Common/interface/RefToBase.h:107:9: note: constraints not satisfied
src/DataFormats/Common/interface/RefToBase.h: In substitution of 'template<class REF>  requires requires{typename REF::value_type;requires(is_same_v<T, typename REF::value_type>) || (is_base_of_v<T, typename REF::value_type>) || (is_base_of_v<typename REF::value_type, T>);} REF edm::RefToBase<testreftobase::Base>::castTo() const [with REF = testreftobase::Base]':
src/DataFormats/Common/test/reftobase_t.cpp:70:45:   required from here
src/DataFormats/Common/interface/RefToBase.h:243:7:   required by the constraints of 'template<class T> template<class REF>  requires requires{typename REF::value_type;requires(is_same_v<T, typename REF::value_type>) || (is_base_of_v<T, typename REF::value_type>) || (is_base_of_v<typename REF::value_type, T>);} REF edm::RefToBase<T>::castTo() const'
src/DataFormats/Common/interface/RefToBase.h:102:16:   in requirements  [with T = testreftobase::Base; REF = edm::Ref<std::vector<testreftobase::BadType, std::allocator<testreftobase::BadType> >, testreftobase::BadType, edm::refhelper::FindUsingAdvance<std::vector<testreftobase::BadType, std::allocator<testreftobase::BadType> >, testreftobase::BadType> >]
src/DataFormats/Common/interface/RefToBase.h:104:18: note: nested requirement '((is_same_v<T, typename REF::value_type> || is_base_of_v<T, typename REF::value_type>) || is_base_of_v<typename REF::value_type, T>)' is not satisfied
  104 |         requires std::is_same_v<T, typename REF::value_type> or std::is_base_of_v<T, typename REF::value_type> or
      |         ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  105 |                      std::is_base_of_v<typename REF::value_type, T>;
      |                      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1plus: note: set '-fconcepts-diagnostics-depth=' to at least 2 for more detail

```

the old error (also without its coloring)
```
src/DataFormats/Common/interface/RefToBase.h: In instantiation of 'typename std::enable_if<std::is_class<_Tp>::value>::type edm::{anonymous}::checkTypeCompatibility() [with T = testreftobase::Base; U = testreftobase::BadType; typename std::enable_if<std::is_class<_Tp>::value>::type = void]':
src/DataFormats/Common/interface/RefToBase.h:266:56:   required from 'REF edm::RefToBase<T>::castTo() const [with REF = edm::Ref<std::vector<testreftobase::BadType> >; T = testreftobase::Base]'
src/DataFormats/Common/test/reftobase_t.cppunit.cc:72:39:   required from here
src/DataFormats/Common/interface/RefToBase.h:236:50: error: static assertion failed: RefToBase::castTo error element types are not related by inheritance
  236 |       static_assert(std::is_base_of<T, U>::value || std::is_base_of<U, T>::value,
      |                                            ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/DataFormats/Common/interface/RefToBase.h:236:50: note: '(((bool)std::integral_constant<bool, false>::value) || ((bool)std::integral_constant<bool, false>::value))' evaluates to false
```

One item to note. While using Visual Studio Code to edit the test, the version with concepts was caught in the IDE as a code that will fail (i.e. underlined in red) while the static_assert version is not caught. So tools can better handle the concepts version over the static_assert.

Reply:
Thanks for the error messages. I find our `static_assert` message more clear, but the message from concepts is not too bad, and probably a person with some level of knowledge what the `RefToBase` is about could figure out the reason for the failure.